packages/custom_lint/lib/src/package_utils.dart

```diff
@@ -115,14 +123,74 @@ Future<PackageConfig?> tryParsePackageConfig(Directory directory) async {
   }
 }
 
+/// Finds the project directory associated with an analysis context root
+///
+/// This is a folder that contains both a `pubspec.yaml` and a `.dart_tool/package_config.json` file.
+/// It is either alongside the analysis_options.yaml file, or in a parent directory.
+Directory findProjectDirectory(
+  Directory directory, {
+  Directory? original,
+  bool missingPackageConfigOkay = false,
+}) {
+  final packageConfigFile = directory.packageConfig;
+  final pubspecFile = directory.pubspec;
+  if (packageConfigFile.existsSync() && pubspecFile.existsSync()) {
+    return directory;
+  }
+  if (pubspecFile.existsSync()) {
```

Comment:
Checking for the existence of a file is a bit expensive. We should only do it once per file.